### PR TITLE
Add D3 Histogram for Time Conversion Histogram

### LINKED ISSUE #4855: Time conversion analysis - new view
This issue is to discuss the frontend implementation of the time conversion analysis view. Full product specs can be found [here](https://docs.google.com/document/d/1S7fwha7cph-79YaPADfFKHILCnPmBgk4rqaMu-gPZ3g/edit?ts=60d23ad3#heading=h.tj3fkkvnb1bq).

Initial wireframe draft below (from @corywatilo's new UI on #4535). Looking for feedback @corywatilo @marcushyett-ph @mariusandra @kpthatsme. The histogram is quite simple that's why it's mocked. On the y-axis we have a regular number scale as we have today. On the x-axis we have each bucket range.

image

### PR DESCRIPTION
**Pull request by alexkim205**
## PR Checklist

- [X] A custom (and reusable ♻️) D3 histogram component that is used in our time conversion display.
~~- [ ] Horizontal and vertical layouts~~
~~- [ ] Visual tweaks and animation~~

### To be addressed in other PR's

- [ ] Group outliers (@Twixes is this built out in the backend right now?)
- [ ] Bin counts and autobinning (related to #4995) 
- [ ] All steps (blocked on #5110 being merged in)
- [ ] Accuracy of data partially blocked on #5116.
- [ ] Unordered steps (I think this is still blocked on backend dev)
- [ ] Conversion window
- [ ] Empty states
- [ ] Need to move d3 logic into kea

It was difficult to mimic the exact look of our histogram [Figma](https://www.figma.com/file/9yWtngNb1AIuf6KmXaEPJA/App-doodles?node-id=498%3A39) mocks using `<LineGraph/>` so I decided to implement a histogram chart with D3. This is done through a custom `useD3` hook that lets me manipulate DOM elements using plain JS.

Part of #4855.

**Before - Line Graph**

Screen Shot 2021-07-13 at 12 21 23 AM

**After - custom D3 graph**

https://user-images.githubusercontent.com/13460330/125537930-80c901d9-d79f-4689-bd0f-d8f81cf77fa0.mov

**Action Header Bar**

Screen Shot 2021-07-13 at 9 36 47 PM

## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests
- [ ] Migrations are safe to run at scale (e.g. PostHog Cloud) – present proof if not obvious
- [ ] New/changed UI is decent on smartphones (viewport width around 360px)

### CODE REVIEW

**Review by clarkus — APPROVED** [review]
These changes look good to me. This might also partially address https://github.com/PostHog/posthog/issues/2656

File: frontend/src/scenes/insights/Histogram/Histogram.tsx [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+                .ticks(y.ticks().length)
+
+            const renderCanvas = (parentNode: D3Selector): D3Selector => {
+                // Get or create svg > g
+                const _svg = getOrCreateEl(parentNode, 'svg > g', () =>
+                    parentNode
+                        .append('svg:svg')
+                        .attr('viewBox', `0 0 ${config.width} ${config.height}`)
+                        .append('svg:g')
+                        .classed(isVertical ? FunnelLayout.vertical : FunnelLayout.horizontal, true)
```
Comment (mariusandra) [inline frontend/src/scenes/insights/Histogram/Histogram.tsx:76]: nit: couldn't this be `.classed(layout, true)` to better match the search `.classed(layout)` below?

File: frontend/src/scenes/insights/Histogram/Histogram.tsx [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+                )
+
+                // if class doesn't exist on svg>g, layout has changed. after we learn this, reset
+                // the layout
+                const layoutChanged = !_svg.classed(layout)
+                _svg.attr('class', null).classed(layout, true)
```
Comment (mariusandra) [inline frontend/src/scenes/insights/Histogram/Histogram.tsx:82]: This is fine for now, but we'll need to figure out a good universal way to track changes between values in logics and d3. There might even be a way to use selectors and/or actions directly here. 🤔

**Review by mariusandra — COMMENTED** [review]
Definitely looks good and agreed with the direction. I'm even able to understand the time scale on these graphs, less so in the "before" group :).

However I tried running it with postgres and got:

```
"Uncaught TypeError: Cannot read property 'bin1' of undefined
```

.. and the screen went white. Seems like this is coming from inside the d3 hook. I'd suggest wrapping the component in a React error boundary to prevent any future bugs there from doing more damage than they should.

There are also obvious console.log statements and PR itself is still as draft, so let me know if/when you want a deeper review... and I'll fire up the old houseclicker.

If the feature doesn't work with Postgres, best to disable it as well.

[This line in this PR](https://github.com/PostHog/posthog/pull/5077/files#diff-92e32501a6b77e6bdff549cb4f721a309c1b5fd3df14615dc6969315b66041dcR311) provides a way to check it easily.

**Comment by paolodamico** [comment]
Will wait for @mariusandra's comments to be addressed as I can't test locally. Also, it's marked as draft so not sure if we jumped the gun here.

From the recording you uploaded, I would suggest considering dropping the two visualization options (left-to-right and top-to-bottom) for this view. I don't see a strong case for having the top-to-bottom and feels like it can introduce cognitive burden to users, one more thing to pay attention to and decide on, quoting Marius quoting Steve K, "Don't make me think". CC @clarkus for thoughts too.

**Comment by mariusandra** [comment]
I'm still not able to test this locally, as I don't have any funnel returning any data now. However the empty view here could also be improved somewhat. At least by aligning the axes.
image

**Comment by mariusandra** [comment]
Also, I tend to agree with Paolo here regarding the two choices. Also, I'd personally be in favour of removing the animations. I think they're more distracting than useful TBH. E.g. in the video the bars are animating, but the timescale changes, so it's not really providing anything visually... except just slowing the chart half a second later.

**Comment by mariusandra** [comment]
Yeah, the animations are more distracting than useful :).

**Comment by clarkus** [comment]
> From the recording you uploaded, I would suggest considering dropping the two visualization options (left-to-right and top-to-bottom) for this view. I don't see a strong case for having the top-to-bottom and feels like it can introduce cognitive burden to users, one more thing to pay attention to and decide on, quoting Marius quoting Steve K, "Don't make me think". CC @clarkus for thoughts too.

Agreed with presenting a single option. I think most often, the left-to-right layout of histograms is most valuable. I also think dropping the animations is a good call for now. They can sometimes make a chart feel like it's performing poorly in terms of rendering time.

**Comment by alexkim205** [comment]
I just wanted to collect preliminary thoughts in this draft to see if I was going in the right direction, so thank you all for the helpful feedback!

re: long date formats

I think this would solve the problem for long date labels in instances of histograms (time conversion only for now) in our dashboards. Perhaps we can have an app-wide date shortening initiative or tackle it by insight type since we're revamping quite a few of our graphs.

re: postgres not working

From what I can see, it doesn't look like the new time conversion api is supported by postgres so definitely makes sense to hide this feature behind the `clickhouseFeatures` value that Sam created.

re: vertical and horizontal layout

Initially I added both layouts so that the time conversion experience would be consistent w/ the funnel steps, but I see that the reason why we added the horizontal layout in the first place (to fit funnels with _many_ steps) doesn't really apply here. I'll keep the dual layout functionality in the generic `<Histogram/>` component since it might be useful in different situations (maybe can apply something similar to a custom bar chart), but remove the option to change layouts in the time conversion view.

re: animations

Got it! functionality > 👁️ 🍬

File: frontend/src/scenes/insights/Histogram/Histogram.tsx (hunk 1 above)
Comment (alexkim205) [reply frontend/src/scenes/insights/Histogram/Histogram.tsx:74]: very true 👍

File: frontend/src/scenes/insights/Histogram/Histogram.tsx (hunk 2 above)
Comment (alexkim205) [reply frontend/src/scenes/insights/Histogram/Histogram.tsx:80]: Yup this was kind of hacky and it'd be nice to put it in some kind of `histogramLogic` or something. I just kept it here bc there wasn't much state being passed around. Maybe we can refactor this later ad-hoc

**Comment by alexkim205** [comment]
> I'm still not able to test this locally, as I don't have any funnel returning any data now. However the empty view here could also be improved somewhat. At least by aligning the axes.

I'll address this in a followup PR dedicated to funnel time conversion empty states

File: frontend/src/scenes/funnels/funnelLogic.ts [hunk 3]
```diff
@@ -106,9 +107,9 @@ export const funnelLogic = kea<funnelLogicType<TimeStepOption>>({
 
     loaders: ({ props, values, actions }) => ({
         results: [
-            [] as FunnelStep[],
+            [],
```
Comment (mariusandra) [inline frontend/src/scenes/funnels/funnelLogic.ts:110]: Why remove this?

**Review by mariusandra — APPROVED** [review]
Looks good! I'm not sure why this one type in `funnelLogic` was removed though. Can you add it back if it was a mistake?

File: frontend/src/scenes/funnels/funnelLogic.ts (hunk 3 above)
Comment (alexkim205) [reply frontend/src/scenes/funnels/funnelLogic.ts:110]: Thanks! looks like i fat fingered something here. I'll add it back

**Review by samwinslow — COMMENTED** [review]
Code LGTM but it's confusing that there's no unit listed for the x-axis. Also, nitpick but it would be nice to see a border-radius on this bar to match the step bars.
 edit: disregard as I was looking at an outdated commit.

**Pull request merged**